File: app/src/main/java/com/qurantv/app/data/repo/KsuHilitesRepository.kt
package com.qurantv.app.data.repo

import com.qurantv.app.data.cache.JsonDiskCache
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Exact per-ayah page positions from quran.ksu.edu.sa (the Ayat reference app).
 *
 * `GET interface.php?ui=pc&do=hilites&mosshaf=<mushaf>&t=28&page=<page>` →
 * `{"<page>": {"<sura>_<aya>": [x, y], ...}}` where x/y are the ayah's start
 * in the page image's NATIVE pixel space. Immutable → cached forever on disk.
 */
class KsuHilitesRepository(
    private val api: com.qurantv.app.data.api.ApiClient,
    private val cache: JsonDiskCache,
    private val json: Json,
) {

    /** An ayah's end position on the page, in the site's data space. */
    data class AyahPosition(val surah: Int, val ayah: Int, val x: Int, val y: Int)

    private data class PagePositions(val ordered: List<AyahPosition>)

    private class PageCache : LinkedHashMap<String, PagePositions>(MAX_PAGES, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, PagePositions>): Boolean =
            size > MAX_PAGES
    }

    private val memory = PageCache()

    /** Page ayah end-positions in document order; null on failure. */
    suspend fun positionsFor(mushaf: String, page: Int): List<AyahPosition>? {
        val key = "${mushaf}_$page"
        memory[key]?.let { return it.ordered }
        return cache.singleFlight(key) {
            val cached = cache.read(JsonDiskCache.KSU_HILITES, key)
            if (cached != null) {
                parse(cached)?.also { memory[key] = it }?.ordered
            } else {
                try {
                    val url = "https://quran.ksu.edu.sa/interface.php?ui=pc&do=hilites&mosshaf=$mushaf&t=28&page=$page"
                    val raw = api.getText(url)
                    parse(raw)?.let { parsed ->
                        runCatching { cache.write(JsonDiskCache.KSU_HILITES, key, raw) }
                        memory[key] = parsed
                        parsed.ordered
                    }
                } catch (e: Exception) {
                    null
                }
            }
        }
    }

    private fun parse(raw: String): PagePositions? {
        val obj = runCatching { json.parseToJsonElement(raw) as? JsonObject }.getOrNull() ?: return null
        val pageObj = obj.values.firstOrNull() as? JsonObject ?: return null
        val ordered = ArrayList<AyahPosition>(pageObj.size)
        pageObj.forEach { (key, value) ->
            val parts = key.split('_')
            if (parts.size != 2) return@forEach
            val surah = parts[0].toIntOrNull() ?: return@forEach
            val ayah = parts[1].toIntOrNull() ?: return@forEach
            if (surah !in 1..114) return@forEach
            val arr = value.jsonArray
            if (arr.size < 2) return@forEach
            val x = arr[0].jsonPrimitive.content.toIntOrNull() ?: return@forEach
            val y = arr[1].jsonPrimitive.content.toIntOrNull() ?: return@forEach
            ordered += AyahPosition(surah, ayah, x, y)
        }
        if (ordered.isEmpty()) return null
        return PagePositions(ordered)
    }

    companion object {
        private const val MAX_PAGES = 12
    }
}

File: web/src/data/repo/KsuHilitesRepository.ts
// Ported 1:1 from app/src/main/java/com/qurantv/app/data/repo/KsuHilitesRepository.kt
// Exact per-ayah page positions from quran.ksu.edu.sa (the Ayat reference app).
// `interface.php?ui=pc&do=hilites&mosshaf=<m>&t=28&page=<p>` →
// `{"<p>": {"<sura>_<aya>": [x, y], ...}}` — immutable, cached forever.

import type { KsuAyahEnd } from "../../domain/KsuHiliteGeometry";
import { CACHE, JsonDiskCache } from "../cache/JsonDiskCache";

const MAX_PAGES = 12;

export class KsuHilitesRepository {
  private readonly memory = new Map<string, KsuAyahEnd[]>();

  constructor(
    private readonly cache: JsonDiskCache,
    private readonly fetchText: (url: string) => Promise<string>,
  ) {}

  /** Page ayah end-positions in document order; null on failure. */
  async positionsFor(mushaf: string, page: number): Promise<KsuAyahEnd[] | null> {
    const key = `${mushaf}_${page}`;
    const mem = this.memory.get(key);
    if (mem) return mem;
    return this.cache.singleFlight(key, async () => {
      const cached = await this.cache.read(CACHE.KSU_HILITES, key);
      if (cached !== null) {
        const parsed = parseHilites(cached);
        if (parsed) this.memoize(key, parsed);
        return parsed;
      }
      try {
        const url = `https://quran.ksu.edu.sa/interface.php?ui=pc&do=hilites&mosshaf=${mushaf}&t=28&page=${page}`;
        const raw = await this.fetchText(url);
        const parsed = parseHilites(raw);
        if (parsed) {
          await this.cache.write(CACHE.KSU_HILITES, key, raw);
          this.memoize(key, parsed);
        }
        return parsed;
      } catch {
        return null;
      }
    });
  }

  private memoize(key: string, positions: KsuAyahEnd[]): void {
    this.memory.set(key, positions);
    if (this.memory.size > MAX_PAGES) {
      const first = this.memory.keys().next().value;
      if (first !== undefined) this.memory.delete(first);
    }
  }
}

export function parseHilites(raw: string): KsuAyahEnd[] | null {
  let obj: unknown;
  try {
    obj = JSON.parse(raw);
  } catch {
    return null;
  }
  if (typeof obj !== "object" || obj === null) return null;
  const pageObj = Object.values(obj)[0];
  if (typeof pageObj !== "object" || pageObj === null) return null;
  const ordered: KsuAyahEnd[] = [];
  for (const [key, value] of Object.entries(pageObj)) {
    const parts = key.split("_");
    if (parts.length !== 2) continue;
    const surah = Number.parseInt(parts[0], 10);
    const ayah = Number.parseInt(parts[1], 10);
    if (!Number.isInteger(surah) || surah < 1 || surah > 114 || !Number.isInteger(ayah)) continue;
    if (!Array.isArray(value) || value.length < 2) continue;
    const x = Number.parseInt(String(value[0]), 10);
    const y = Number.parseInt(String(value[1]), 10);
    if (!Number.isInteger(x) || !Number.isInteger(y)) continue;
    ordered.push({ surah, ayah, x, y });
  }
  return ordered.length > 0 ? ordered : null;
}
